### PR TITLE
updated max connections to max agent

### PR DESCRIPTION
**Pull request by lykkin**
## What does this PR do?

Converts the max connections to a max agents field. This change is related to https://github.com/elastic/fleet-server/pull/1014.

Technically a breaking change, since this removes a field that may be used by users. 

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

Host the package registry, enroll and run an agent, set a value in the Max agents field and observe the config come down as such:

```
...
    inputs:
    - data_stream:
        namespace: default
      id: default-fleet-server-agent-policy
      meta:
        package:
          name: fleet_server
          version: 2.0.0
      name: fleet_server-1
      revision: 3
      server:
        host: 0.0.0.0
        limits.max_agents: 100
        port: 8220
      type: fleet-server
      use_output: default

...
```

## Related issues

Closes https://github.com/elastic/fleet-server/issues/571

### CODE REVIEW

**Review by ruflin — CHANGES_REQUESTED** [review]
I would like to better understand what is the migration path for our users here? What if users have set the setting in a previous version and upgrade? Do we need to adjust the compatibility of this package to a more recent version of Kibana?

So far we tried to stay away from breaking changes in packages and especially for a package like fleet_server that every Elastic Agent users use, we should not introduce any change that will break users just upgrading.

Lets figure out the migration story. If there is already a plan, can you link it here to read up?

**Review by ph — CHANGES_REQUESTED** [review]
See issue with goroutine.

**Comment by lykkin** [comment]
@ruflin @nimarezainia i started [a migration doc](https://docs.google.com/document/d/1oeCvQ-QXUU1d7KQeyXPTR_kex8zmDtR3viUwBqwDyRU/edit) where we can start talking about how to handle the update.

**Comment by jlind23** [comment]
@lykkin any update there?

**Comment by jlind23** [comment]
@lykkin is it ready to be reviewed again?

File: packages/fleet_server/manifest.yml [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
+version: 1.2.0
 release: ga
 description: Centrally manage Elastic Agents with the Fleet Server integration.
 type: integration
 format_version: 1.0.0
 license: basic
 categories: ["elastic_stack"]
 conditions:
-  kibana.version: "^7.16.0 || ^8.0.0"
+  kibana.version: "^7.16.0 || ^8.1.0"
```
Comment (ruflin) [inline packages/fleet_server/manifest.yml:11]: It seems odd that someone can install a package in 7.16, but then when going to 8.0 it is not compatible anymore and only compatible in 8.1?

**Review by ruflin — COMMENTED** [review]
What is the migration story we landed on? Lets add the details to this Pull Request so everyone is aware and the PR can be reviewed with this migration story in mind.

File: packages/fleet_server/manifest.yml (hunk 1 above)
Comment (lykkin) [reply packages/fleet_server/manifest.yml:11]: good catch, this was left over from when we were going to land UI changes alongside this.

**Comment by lykkin** [comment]
> migration story we landed on? Lets add the details to this Pull Request so everyone is aware and the PR can be reviewed with this migration story in mind.

we are going to update the UI to have the new field and handle reconciliation of the values on the fleet server side. there will be some UI changes to make it easier to work with in a later release.

**Comment by ruflin** [comment]
> Technically a breaking change, since this removes a field that may be used by users.
Nit: Can you cleanup the PR description. I assume this is not true anymore?

In this PR the required Kibana version is not changed. But https://github.com/elastic/fleet-server/pull/1014 is not merged yet and very unlikely backported to 7.x. What happens if this is used with an 8.0 or 7.last fleet-server version that does not support this config yet. There is also the scenario, that Kibana is actually on a more recent version but fleet-server is not upgraded yet.

**Comment by lykkin** [comment]
> > Technically a breaking change, since this removes a field that may be used by users.
> > Nit: Can you cleanup the PR description. I assume this is not true anymore?
> 
> In this PR the required Kibana version is not changed. But [elastic/fleet-server#1014](https://github.com/elastic/fleet-server/pull/1014) is not merged yet and very unlikely backported to 7.x. What happens if this is used with an 8.0 or 7.last fleet-server version that does not support this config yet. There is also the scenario, that Kibana is actually on a more recent version but fleet-server is not upgraded yet.

In those cases the field will be ignored by fleet-server, however there will be no feedback to the user that this will be an issue.

Is there a way to add a condition for this case? If not, do we report library version to ES?

I'm not entirely sure what the best way forward here is. @ruflin @ph, thoughts?

**Comment by ph** [comment]
I've given some thoughts on this, I do not consider this a breaking change because of the following:

- We are actually giving them a better experience at the cost of removing an option.
- We aren't crashing hard on the use.

Now concerning the behavior of the Fleet Server, we should make sure we log that you are using a deprecated field.

@ruflin Concerning the case of running an 8.0 Kibana and a 7.last Fleet-Server, I am not sure this really applies. Because 8.0 needs to support 7. last. I am not sure that 8.2 is a hard requirement to support 7.last and we do suggest upgrading and have the stack aligned.

**Review by ph — APPROVED** [review]

**Comment by lykkin** [comment]
We are actually giving them a better experience at the cost of removing an option.

Currently we aren't removing a field, just deprecating it. We will be handling the reconciliation of the values when they hit fleet-server. This _should_ avoid a breaking change, as it won't change behavior until they update their configs to use the new value.

**Comment by ruflin** [comment]
Lets assume we have a user that uses fleet-server 8.0 and Kibana 8.3. That user will put in a config option assuming it works. But in reality, nothing happens. Maybe we can solve this by just adding a note in the description of the config option?

**Comment by lykkin** [comment]
> Lets assume we have a user that uses fleet-server 8.0 and Kibana 8.3. That user will put in a config option assuming it works. But in reality, nothing happens. Maybe we can solve this by just adding a note in the description of the config option?

 I think this is a decent compromise. Ideally, I think we would want something programmatic that would gate fields based on installed library versions, but I think that would be a future feature.

File: packages/fleet_server/manifest.yml [hunk 2]
```diff
@@ -40,11 +40,22 @@ policy_templates:
             show_user: true
             default:
               - 8220
+          - name: max_agents
+            type: integer
+            title: Max agents (per server)
+            required: false
+            show_user: true
+            description: >
+              Requires fleet-server v8.2 and higher.
```
Comment (lykkin) [inline packages/fleet_server/manifest.yml:49]: I'll update this to be 8.3 if we decide to punt due to feature freeze

**Comment by jlind23** [comment]
@ruflin can we get your eyes on this as you are the last step before being able to merge.

**Review by ruflin — APPROVED** [review]
LGTM. I assume this was tested with 8.3 but also 8.0 / 7.17 version of fleet-server. In theory it should work, but lets make sure we do not break any old deployments when this package is upgraded.

**Comment by jlind23** [comment]
@ph can you merge it please?

**Pull request merged**